kornia/geometry/depth.py

```diff
@@ -2,23 +2,115 @@
+) -> Tensor:
+    """Compute a 3d point per pixel given its depth value and the camera intrinsics.
+
+    .. tip::
+
+        This function should be used in conjunction with :py:func:`kornia.geometry.depth.depth_to_3d_v2` to cache
+        the meshgrid computation when warping multiple frames with the same camera intrinsics.
+
+    Args:
+        camera_matrix: tensor containing the camera intrinsics with shape :math:`(3, 3)`.
+        normalize_points: whether to normalise the pointcloud. This must be set to `True` when the depth is
+          represented as the Euclidean ray length from the camera position.
+
+    Return:
+        tensor with a 3d point per pixel of the same resolution as the input :math:`(*, H, W, 3)`.
+    """
+    KORNIA_CHECK_SHAPE(camera_matrix, ["3", "3"])
+
+    # create base coordinates grid
+    points_uv: Tensor = create_meshgrid(
+        height, width, normalized_coordinates=False, device=device, dtype=dtype
+    ).squeeze()  # HxWx2
+
+    points_xy = normalize_points_with_intrinsics(points_uv, camera_matrix)  # HxWx2
+
+    # unproject pixels to camera frame
+    points_xyz = convert_points_to_homogeneous(points_xy)  # HxWx3
+
+    if normalize_points:
+        points_xyz = kornia_ops.normalize(points_xyz, dim=-1, p=2)
+
+    return points_xyz
+
+
+def depth_to_3d_v2(
+    depth: Tensor, camera_matrix: Tensor, normalize_points: bool = False, xyz_grid: Tensor | None = None
+) -> Tensor:
+    """Compute a 3d point per pixel given its depth value and the camera intrinsics.
+
+    .. note::
+
+        This is an alternative implementation of :py:func:`kornia.geometry.depth.depth_to_3d`
+        that does not require the creation of a meshgrid.
+
+    Args:
+        depth: image tensor containing a depth value per pixel with shape :math:`(*, H, W)`.
+        camera_matrix: tensor containing the camera intrinsics with shape :math:`(*, 3, 3)`.
+        normalize_points: whether to normalise the pointcloud. This must be set to `True` when the depth is
+          represented as the Euclidean ray length from the camera position.
+
+    Return:
+        tensor with a 3d point per pixel of the same resolution as the input :math:`(*, H, W, 3)`.
+
+    Example:
+        >>> depth = torch.rand(4, 4)
+        >>> K = torch.eye(3)
+        >>> depth_to_3d(depth, K).shape
+        torch.Size([4, 4, 3])
+    """
+    KORNIA_CHECK_SHAPE(depth, ["*", "H", "W"])
+    KORNIA_CHECK_SHAPE(camera_matrix, ["*", "3", "3"])
+
+    # create base grid if not provided
+    points_xyz = xyz_grid
+    if xyz_grid is None:
+        height, width = depth.shape[-2:]
+        points_xyz = unproject_meshgrid(height, width, camera_matrix, normalize_points, depth.device, depth.dtype)
+
+    KORNIA_CHECK_SHAPE(points_xyz, ["*", "H", "W", "3"])
+
+    return points_xyz * depth[..., None]  # HxWx3
+
+
+# NOTE: this function should replace the old depth_to_3d
```

Comment:
we should add the deprecated decorator there: https://github.com/kornia/kornia/blob/a212b4d0746b15a4e56a34e72ddee9b320284fb9/kornia/utils/helpers.py#L70